torch/csrc/jit/scope.cpp

```diff
@@ -77,5 +78,50 @@ std::string Scope::namesFromRoot(const std::string& separator) const {
   return out;
 }
 
+InlinedCallStackPtr InlinedCallStack::intrusive_from_this() {
+  c10::raw::intrusive_ptr::incref(this); // we are creating a new pointer
+                                         // from a raw `this` pointer
+                                         // so we need to bump the refcount
+                                         // to account for this ownership
+  return c10::intrusive_ptr<InlinedCallStack>::reclaim(this);
+}
+
+InlinedCallStack::InlinedCallStack(Function* fn, SourceRange source_range)
+    : fn_(fn), source_range_(std::move(source_range)) {}
+
+InlinedCallStack::InlinedCallStack(
+    InlinedCallStackPtr caller,
+    Function* fn,
+    SourceRange source_range)
+    : caller_(std::move(caller)),
+      fn_(fn),
+      source_range_(std::move(source_range)) {}
+
+InlinedCallStackPtr InlinedCallStack::insertCallStackEntry(
+    Function* fn,
+    const SourceRange& source_range) {
+  auto ent = std::make_pair(fn, source_range);
+  if (callees_.count(ent)) {
+    return callees_.at(ent);
+  }
+  auto subscope = c10::make_intrusive<InlinedCallStack>(
+      intrusive_from_this(), fn, source_range);
+  callees_[ent] = subscope;
+  return subscope;
+}
+
+c10::optional<InlinedCallStackPtr> InlinedCallStack::caller() const {
+  return caller_;
+}
+
+std::vector<InlinedCallStackEntry> InlinedCallStack::vec() {
+  std::vector<InlinedCallStackEntry> r;
+  c10::optional<InlinedCallStackPtr> current = intrusive_from_this();
+  while (current) {
+    r.emplace_back(std::make_pair((*current)->fn_, (*current)->source_range_));
+    current = (*current)->caller_;
```

Comment:
Something is confusing to me here. The _last_ thing in this list is the inner-most function, right? The docs say that and the tests also test that. But the inner-most thing cannot be the "caller" of anything. Nevertheless, if it is the last thing in this loop, current->caller_ must at some point point to this inner-most thing. I believe that the `caller_` `callee_` names are swapped in the Trie. My follow up thought is that I don't understand how the Trie is memoizing anything. The inner most link is the most inlined node, but this node will only appear once. Is it actually caching anything?

Reply:
You were correct, the caching didn't work properly. I fixed that and added a test to verify that we're actually sharing callstack-entries among different nodes when it's possible. Please take another look!